File: src/main/java/dev/toastmc/toastclient/api/util/ToastColor.java
package dev.toastmc.toastclient.api.util;

import com.mojang.blaze3d.platform.GlStateManager;

import java.awt.*;

public class ToastColor extends java.awt.Color {

    public static ToastColor rainbow() {
        float hue = (System.currentTimeMillis() % (320 * 32)) / (320f * 32);
        return new ToastColor(ToastColor.fromHSB(hue, 1, 1, 255));
    }

    private static final long serialVersionUID = 1L;

    public ToastColor(int rgb) {
        super(rgb);
    }

    public ToastColor(int rgba, boolean alpha) {
        super(rgba,alpha);
    }

    public ToastColor(int r, int g, int b) {
        super(r,g,b);
    }

    public ToastColor(int r, int g, int b, int a) {
        super(r,g,b,a);
    }

    public ToastColor(java.awt.Color color) {
        super(color.getRed(),color.getGreen(),color.getBlue(),color.getAlpha());
    }

    public ToastColor(ToastColor color, int a) {
        super(color.getRed(),color.getGreen(),color.getBlue(),a);
    }

    public static ToastColor fromHSB(float hue, float saturation, float brightness, int alpha) {
        Color hsbColor = java.awt.Color.getHSBColor(hue,saturation,brightness);
        return new ToastColor(hsbColor.getRed(), hsbColor.getGreen(), hsbColor.getBlue(), alpha);
    }

    public float getHue() {
        return RGBtoHSB(getRed(),getGreen(),getBlue(),null)[0];
    }

    public float getSaturation() {
        return RGBtoHSB(getRed(),getGreen(),getBlue(),null)[1];
    }

    public float getBrightness() {
        return RGBtoHSB(getRed(),getGreen(),getBlue(),null)[2];
    }

    public int getABGRPackedInt() {
        int i = -1;
        try {
            i = Integer.parseInt("0x" + Integer.toHexString(getRGB()));
        } catch (NumberFormatException ignored) { }
        return i;
    }

    public void glColor() {
        GlStateManager.color4f(getRed()/255.0f,getGreen()/255.0f,getBlue()/255.0f,getAlpha()/255.0f);
    }
}

File: src/main/kotlin/dev/toastmc/toastclient/api/util/ToastColor.kt
package dev.toastmc.toastclient.api.util

import com.mojang.blaze3d.systems.RenderSystem
import java.awt.Color

class ToastColor : Color {
    constructor(rgb: Int) : super(rgb) {}
    constructor(rgba: Int, alpha: Boolean) : super(rgba, alpha) {}
    constructor(r: Int, g: Int, b: Int) : super(r, g, b) {}
    constructor(r: Int, g: Int, b: Int, a: Int) : super(r, g, b, a) {}
    constructor(color: Color) : super(color.red, color.green, color.blue, color.alpha) {}
    constructor(color: ToastColor, a: Int) : super(color.red, color.green, color.blue, a) {}

    companion object {
        private const val serialVersionUID = 1L

        fun rainbow(alpha: Int): ToastColor {
            val hue = System.currentTimeMillis() % (320 * 32) / (320f * 32)
            return ToastColor(fromHSB(hue, 1f, 1f, alpha))
        }

        fun fromHSB(hue: Float, saturation: Float, brightness: Float, alpha: Int): ToastColor {
            val hsbColor = getHSBColor(hue, saturation, brightness)
            return ToastColor(hsbColor.red, hsbColor.green, hsbColor.blue, alpha)
        }
    }

    val hue: Float
        get() = RGBtoHSB(red, green, blue, null)[0]
    val saturation: Float
        get() = RGBtoHSB(red, green, blue, null)[1]
    val brightness: Float
        get() = RGBtoHSB(red, green, blue, null)[2]

    val aBGRPackedInt: Int
        get() {
            var i = -1
            try {
                i = ("0x" + Integer.toHexString(rgb)).toInt()
            } catch (ignored: NumberFormatException) {

            }
            return i
        }

    fun glColor() {
        RenderSystem.setShaderColor(red / 255.0f, green / 255.0f, blue / 255.0f, alpha / 255.0f)
    }

    fun glClear() {
        RenderSystem.clearColor(red / 255.0f, green / 255.0f, blue / 255.0f, alpha / 255.0f)
    }
}